openedx-unsupported/ecommerce
# ENT-169: Update offer's Range model to use course catalog query

…nterprise coupon

ENT-169
@asadiqbal08 @saleem-latif @mattdrayer 

* Update the method `contains_product` of `Range` model to get the catalog query from the linked catalog in the course catalog service.
* Update the method `run_catalog_query` of `Range` model to use optional parameter `query` for custom catalog course run queries (in case of course catalog coupons)
* Add method `is_course_in_enterprise_courses` for verification that the basket product (course run) exists in the learner's related enterprise catalog.
* Remove the custom entitlement method `get_course_ids_from_voucher` for enterprise and use the updated method `contains_product` of `Range` model for checking the product in the enterprise entitlement/coupon course runs
* Fix bug [ENT-174](https://openedx.atlassian.net/browse/ENT-174): For enterprise entitlements, first verify that the basket product (course run) exists in the related enterprise catalog before checking the product in the related coupon course runs list
* Update auto entitlements to be able to use multiple coupon codes
* Add/Update unittests

## Testing Steps:
**Sandbox:** https://zubair-arbi.sandbox.edx.org/

As a staff user setup an enterprise with entitlements:
1. Add a course catalog in Discovery django admin: https://discovery-enterprise.sandbox.edx.org/admin/catalogs/catalog/
2. Add an enterprise using LMS django admin with course catalog created above: https://zubair-arbi.sandbox.edx.org/admin/enterprise/enterprisecustomer/
3. Use CAT tool in ECOM to add a catalog coupon (enrollment or discount) with with enterprise course catalog: https://ecommerce-zubair-arbi.sandbox.edx.org/coupons/
4. Add the newly created coupon id (get it from the coupon url, e.g. 65 in "https://ecommerce-zubair-arbi.sandbox.edx.org/coupons/65/") and add it in the section "Enterprise Customer Entitlements" in LMS django admin for enterprise.
5. Link some user/learner with the enterprise using LMS django admin: https://zubair-arbi.sandbox.edx.org/admin/enterprise/enterprisecustomeruser/add/

As an enterprise learner:
1. Try to enroll in the course available in the enterprise catalog and verify that the enterprise entitlement coupon voucher is automatically applied.
2. Remove the auto applied enterprise coupon.
3. Try to enroll in the course which is not available in the enterprise catalog and verify that the enterprise entitlement coupon voucher is not applied.

_**Note:** For testing you can use the enterprise learner "`zubair@edx.org`" and an existing course "https://zubair-arbi.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/about" which are linked with the enterprise "`Arbisoft ENT`" which has an enrollment discount coupon "https://ecommerce-zubair-arbi.sandbox.edx.org/coupons/65" available._
